primap2/csg/_strategies/global_least_squares.py

```diff
@@ -0,0 +1,267 @@
+import numpy as np
+import xarray as xr
+from attrs import frozen
+from scipy.linalg import lstsq
+from scipy.optimize import least_squares
+
+import primap2
+
+
+@frozen
+class GlobalLSStrategy:
+    """Fill missing data by global least square matching.
+
+    The NaNs in the first timeseries `ts(t)` using harmonized data from the lower
+    priority timeseries `fill_ts(t)`. For harmonization we use
+    fill_ts(t)_h = ts(t) * a + b,
+    where fill_ts(t)_h is the harmonized dataset and a and b are determined by minimizing
+    the least squares distance between ts(t) and fill_ts(t)_h.
+
+    If the class is initialized with `allow_shift = True` the faster
+    `GlobalLSlstsqStrategy` which is based on the `scipy.linalg.lstsq` function
+    is used.
+    For the case `allow_shift = False` (b = 0) `scipi.optimize.least_squares` is used.
+
+    If there is no overlap in non-NaN data between ts(t) and fill_ts(t) the
+    `SubstitutionStrategy` is used.
+    """
+
+    allow_shift: bool = True
+    type = "globalLS"
+
+    def factor_mult(self, a, e, e_ref):
+        return a * e - e_ref
+
+    def jac(self, a, e, e_ref):
+        J = np.empty((e.size, 1))
+        J[:, 0] = e
+        return J
+
+    def fill(
+        self,
+        *,
+        ts: xr.DataArray,
+        fill_ts: xr.DataArray,
+        fill_ts_repr: str,
+    ) -> tuple[xr.DataArray, list[primap2.ProcessingStepDescription]]:
+        """Fill missing data by global least square matching.
+
+        The NaNs in the first timeseries `ts(t)` using harmonized data from the
+        lower priority timeseries `fill_ts(t)`. For harmonization we use
+        fill_ts(t)_h = ts(t) * a + b,
+        where fill_ts(t)_h is the harmonized dataset and a and b are determined
+        by minimizing the least squares distance between ts(t) and fill_ts(t)_h.
+
+        If the class is initialized with `allow_shift = True` the faster
+        `GlobalLSlstsqStrategy` which is based on the `scipy.linalg.lstsq`
+        function is used.
+        For the case `allow_shift = False` (b = 0) `scipi.optimize.least_squares`
+        is used.
+
+        If there is no overlap in non-NaN data between ts(t) and fill_ts(t) the
+        `SubstitutionStrategy` is used.
+
```

Comment:
Probably not very useful to repeat the documentation from the class docstring. Maybe:
```suggestion

    For a description of the algorithm, see the documentation of this class.

```